File: backend/repositories/jobRepository.js
// backend\repositories\jobRepository.js
// ============================================================================
// GENERAL CRUD
// ============================================================================

export const create = async (
  connection,
  { userId, jobType, filename, filePath, notes, options }
) => {
  const optionsStr = options ? JSON.stringify(options) : null;
  const [result] = await connection.query(
    `INSERT INTO import_jobs (user_id, job_type, original_filename, file_path, status, notes, options, created_at, updated_at)
      VALUES (?, ?, ?, ?, 'PENDING', ?, ?, NOW(), NOW())`,
    [userId, jobType, filename, filePath, notes || null, optionsStr]
  );
  return result.insertId;
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} jobId
 * @param {Object} options
 * @returns {Promise<any>}
 */
export const update = async (connection, jobId, { status, summary, errorLog }) => {
  const errorLogStr = errorLog ? JSON.stringify(errorLog) : null;
  if (status === "PROCESSING") {
    await connection.query(
      `UPDATE import_jobs
        SET status = ?, log_summary = ?, error_log = ?, processing_started_at = NOW(), updated_at = NOW()
        WHERE id = ?`,
      [status, summary, errorLogStr, jobId]
    );
  } else {
    await connection.query(
      `UPDATE import_jobs
        SET status = ?, log_summary = ?, error_log = ?, updated_at = NOW()
        WHERE id = ?`,
      [status, summary, errorLogStr, jobId]
    );
  }
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} userId
 * @param {number} limit
 * @returns {Promise<any>}
 */
export const findByUser = async (connection, userId, limit = 20, jobTypes = null) => {
  let query = `
     SELECT ij.*, u.username as uploader_name 
     FROM import_jobs ij
     LEFT JOIN users u ON ij.user_id = u.id
     WHERE ij.user_id = ? 
  `;
  const params = [userId];

  if (jobTypes && Array.isArray(jobTypes) && jobTypes.length > 0) {
    query += ` AND ij.job_type IN (?) `;
    params.push(jobTypes);
  }

  query += ` ORDER BY ij.created_at DESC LIMIT ?`;
  params.push(limit);

  const [rows] = await connection.query(query, params);
  return rows;
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number} limit
 * @param {string[]|null} jobTypes
 * @returns {Promise<any>}
 */
export const findAll = async (connection, limit = 50, jobTypes = null) => {
  let query = `
     SELECT ij.*, u.username as uploader_name 
     FROM import_jobs ij
     LEFT JOIN users u ON ij.user_id = u.id
  `;
  const params = [];

  if (jobTypes && Array.isArray(jobTypes) && jobTypes.length > 0) {
    query += ` WHERE ij.job_type IN (?) `;
    params.push(jobTypes);
  }

  query += ` ORDER BY ij.created_at DESC LIMIT ?`;
  params.push(limit);

  const [rows] = await connection.query(query, params);
  return rows;
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} jobId
 * @param {any} processed
 * @param {any} total
 * @returns {Promise<any>}
 */
export const updateProgress = async (connection, jobId, processed, total) => {
  return connection.query(
    `UPDATE import_jobs
      SET processed_records = ?, total_records = ?, updated_at = NOW()
      WHERE id = ?`,
    [processed, total, jobId]
  );
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} jobId
 * @param {number|string} userId
 * @returns {Promise<any>}
 */
export const voidJob = async (connection, jobId, userId) => {
  const [result] = await connection.query(
    `UPDATE import_jobs
      SET status = 'VOID', updated_at = NOW()
      WHERE id = ? AND status = 'PENDING' AND user_id = ?`,
    [jobId, userId]
  );
  return result.affectedRows > 0;
};

// ============================================================================
// IMPORT JOBS (Worker Specific)
// ============================================================================

/**
 * Ambil Job Pending ATAU Job Retrying yang sudah menunggu > 10 detik secara atomik
 * Mencegah Race Condition antar cron.
 */
export const getAndLockPendingImportJob = async (connection) => {
  await connection.beginTransaction();
  try {
    const [rows] = await connection.query(
      `SELECT * FROM import_jobs
        WHERE status = 'PENDING'
        OR (status = 'RETRYING' AND updated_at <= NOW() - INTERVAL 10 SECOND)
        ORDER BY created_at ASC
        LIMIT 1
        FOR UPDATE`
    );

    if (rows.length === 0) {
      await connection.commit();
      return null;
    }

    const job = rows[0];

    await connection.query(
      `UPDATE import_jobs SET status = 'PROCESSING', processing_started_at = NOW() WHERE id = ?`,
      [job.id]
    );

    await connection.commit();
    return job;
  } catch (error) {
    await connection.rollback();
    throw error;
  }
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} jobId
 * @param {any} currentRetryCount
 * @param {any} errorMessage
 * @returns {Promise<any>}
 */
export const retryImportJob = async (connection, jobId, currentRetryCount, errorMessage) => {
  const nextRetry = currentRetryCount + 1;
  const note = `Retry #${nextRetry}: ${errorMessage.substring(0, 100)}...`;

  return connection.query(
    `UPDATE import_jobs
      SET status = 'RETRYING',
        retry_count = ?,
        log_summary = IF(log_summary IS NULL, ?, CONCAT(log_summary, ' | ', ?)),
        updated_at = NOW()
      WHERE id = ?`,
    [nextRetry, note, note, jobId]
  );
};

// ... (Fungsi complete, fail, export lainnya TETAP SAMA) ...
export const completeImportJob = async (
  connection,
  jobId,
  status,
  summary,
  errorLogJSON = null
) => {
  return connection.query(
    `UPDATE import_jobs
      SET status = ?, log_summary = ?, error_log = ?, updated_at = NOW()
      WHERE id = ?`,
    [status, summary, errorLogJSON, jobId]
  );
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} jobId
 * @param {any} summary
 * @returns {Promise<any>}
 */
export const failImportJob = async (connection, jobId, summary) => {
  return connection.query(
    `UPDATE import_jobs SET status = 'FAILED', log_summary = ?, updated_at = NOW() WHERE id = ?`,
    [summary, jobId]
  );
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {any} timeoutMinutes
 * @returns {Promise<any>}
 */
export const timeoutStuckImportJobs = async (connection, timeoutMinutes) => {
  return connection.query(
    `UPDATE import_jobs
      SET status = 'FAILED',
          log_summary = CONCAT(COALESCE(log_summary, ''), ' [SYSTEM: Job Killed due to timeout/crash]'),
          updated_at = NOW()
      WHERE status = 'PROCESSING'
      AND updated_at < NOW() - INTERVAL ? MINUTE`,
    [timeoutMinutes]
  );
};

// ============================================================================
// EXPORT JOBS
// ============================================================================

// Fungsi untuk membuat Job Export baru
/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {Object} options
 * @returns {Promise<any>}
 */
export const createExportJob = async (connection, { userId, filters, jobType }) => {
  const filtersStr = JSON.stringify(filters);
  const [result] = await connection.query(
    `INSERT INTO export_jobs (user_id, status, filters, job_type, created_at)
      VALUES (?, 'PENDING', ?, ?, NOW())`,
    [userId, filtersStr, jobType || "STOCK_REPORT"]
  );
  return result.insertId;
};

/**
 * Ambil dan lock Job Export secara atomik
 * @param {import('mysql2/promise').Connection} connection
 * @returns {Promise<any>}
 */
export const getAndLockPendingExportJob = async (connection) => {
  await connection.beginTransaction();
  try {
    const [rows] = await connection.query(
      `SELECT * FROM export_jobs WHERE status = 'PENDING' ORDER BY created_at ASC LIMIT 1 FOR UPDATE`
    );

    if (rows.length === 0) {
      await connection.commit();
      return null;
    }

    const job = rows[0];

    await connection.query(
      `UPDATE export_jobs SET status = 'PROCESSING', processing_started_at = NOW() WHERE id = ?`,
      [job.id]
    );

    await connection.commit();
    return job;
  } catch (error) {
    await connection.rollback();
    throw error;
  }
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} jobId
 * @param {any} filename
 * @returns {Promise<any>}
 */
export const completeExportJob = async (connection, jobId, filename) => {
  return connection.query(
    `UPDATE export_jobs SET status = 'COMPLETED', file_path = ? WHERE id = ?`,
    [filename, jobId]
  );
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {number|string} jobId
 * @param {any} errorMessage
 * @returns {Promise<any>}
 */
export const failExportJob = async (connection, jobId, errorMessage) => {
  return connection.query(
    `UPDATE export_jobs SET status = 'FAILED', error_message = ? WHERE id = ?`,
    [errorMessage, jobId]
  );
};

/**
 * @param {import('mysql2/promise').Connection} connection
 * @param {any} timeoutMinutes
 * @returns {Promise<any>}
 */
export const timeoutStuckExportJobs = async (connection, timeoutMinutes) => {
  return connection.query(
    `UPDATE export_jobs
      SET status = 'FAILED', error_message = CONCAT('Job timeout after ', ?, ' minutes')
      WHERE status = 'PROCESSING'
      AND processing_started_at < NOW() - INTERVAL ? MINUTE`,
    [timeoutMinutes, timeoutMinutes]
  );
};

File: backend/internal/repository/job_repository.go
package repository

import (
	"context"

	"github.com/dps-wmhris/backend/internal/model"
	"github.com/jmoiron/sqlx"
)

type JobRepository interface {
	// Import Jobs
	CreateImportJob(ctx context.Context, job *model.ImportJob) (int, error)
	GetImportJobs(ctx context.Context, limit int, offset int) ([]model.ImportJob, error)
	GetImportJobByID(ctx context.Context, id int) (*model.ImportJob, error)
	UpdateImportJobStatus(ctx context.Context, id int, status string, logSummary *string, errorLog *string) error
	GetPendingImportJobs(ctx context.Context) ([]model.ImportJob, error)

	// Export Jobs
	CreateExportJob(ctx context.Context, job *model.ExportJob) (int, error)
	GetExportJobs(ctx context.Context, limit int, offset int) ([]model.ExportJob, error)
	GetExportJobByID(ctx context.Context, id int) (*model.ExportJob, error)
	UpdateExportJobStatus(ctx context.Context, id int, status string, filePath *string, errorMessage *string) error
	GetPendingExportJobs(ctx context.Context) ([]model.ExportJob, error)
}

type jobRepositoryImpl struct {
	db *sqlx.DB
}

func NewJobRepository(db *sqlx.DB) JobRepository {
	return &jobRepositoryImpl{db: db}
}

// --- Import Jobs ---

func (r *jobRepositoryImpl) CreateImportJob(ctx context.Context, job *model.ImportJob) (int, error) {
	query := `
		INSERT INTO import_jobs (user_id, job_type, original_filename, file_path, notes, options, status)
		VALUES (?, ?, ?, ?, ?, ?, 'PENDING')
	`
	res, err := r.db.ExecContext(ctx, query, job.UserID, job.JobType, job.OriginalFilename, job.FilePath, job.Notes, job.Options)
	if err != nil {
		return 0, err
	}
	id, err := res.LastInsertId()
	return int(id), err
}

func (r *jobRepositoryImpl) GetImportJobs(ctx context.Context, limit int, offset int) ([]model.ImportJob, error) {
	query := `SELECT * FROM import_jobs ORDER BY created_at DESC LIMIT ? OFFSET ?`
	var jobs []model.ImportJob
	err := r.db.SelectContext(ctx, &jobs, query, limit, offset)
	return jobs, err
}

func (r *jobRepositoryImpl) GetImportJobByID(ctx context.Context, id int) (*model.ImportJob, error) {
	query := `SELECT * FROM import_jobs WHERE id = ?`
	var job model.ImportJob
	err := r.db.GetContext(ctx, &job, query, id)
	return &job, err
}

func (r *jobRepositoryImpl) UpdateImportJobStatus(ctx context.Context, id int, status string, logSummary *string, errorLog *string) error {
	query := `
		UPDATE import_jobs 
		SET status = ?, log_summary = ?, error_log = ?, updated_at = CURRENT_TIMESTAMP
		WHERE id = ?
	`
	_, err := r.db.ExecContext(ctx, query, status, logSummary, errorLog, id)
	return err
}

func (r *jobRepositoryImpl) GetPendingImportJobs(ctx context.Context) ([]model.ImportJob, error) {
	query := `SELECT * FROM import_jobs WHERE status = 'PENDING' ORDER BY created_at ASC`
	var jobs []model.ImportJob
	err := r.db.SelectContext(ctx, &jobs, query)
	return jobs, err
}

// --- Export Jobs ---

func (r *jobRepositoryImpl) CreateExportJob(ctx context.Context, job *model.ExportJob) (int, error) {
	query := `
		INSERT INTO export_jobs (user_id, job_type, filters, status)
		VALUES (?, ?, ?, 'PENDING')
	`
	res, err := r.db.ExecContext(ctx, query, job.UserID, job.JobType, job.Filters)
	if err != nil {
		return 0, err
	}
	id, err := res.LastInsertId()
	return int(id), err
}

func (r *jobRepositoryImpl) GetExportJobs(ctx context.Context, limit int, offset int) ([]model.ExportJob, error) {
	query := `SELECT * FROM export_jobs ORDER BY created_at DESC LIMIT ? OFFSET ?`
	var jobs []model.ExportJob
	err := r.db.SelectContext(ctx, &jobs, query, limit, offset)
	return jobs, err
}

func (r *jobRepositoryImpl) GetExportJobByID(ctx context.Context, id int) (*model.ExportJob, error) {
	query := `SELECT * FROM export_jobs WHERE id = ?`
	var job model.ExportJob
	err := r.db.GetContext(ctx, &job, query, id)
	return &job, err
}

func (r *jobRepositoryImpl) UpdateExportJobStatus(ctx context.Context, id int, status string, filePath *string, errorMessage *string) error {
	query := `
		UPDATE export_jobs 
		SET status = ?, file_path = ?, error_message = ?
		WHERE id = ?
	`
	_, err := r.db.ExecContext(ctx, query, status, filePath, errorMessage, id)
	return err
}

func (r *jobRepositoryImpl) GetPendingExportJobs(ctx context.Context) ([]model.ExportJob, error) {
	query := `SELECT * FROM export_jobs WHERE status = 'PENDING' ORDER BY created_at ASC`
	var jobs []model.ExportJob
	err := r.db.SelectContext(ctx, &jobs, query)
	return jobs, err
}
